### PR TITLE
Syslog support in Vector

### PR DESCRIPTION
**Pull request by syedriko**
### Description
Syslog support in Vector

/cc @jcantrill 
/assign @vimalk78 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-2288

### CODE REVIEW

File: test/framework/e2e/syslog.go [hunk 1]
```diff
@@ -298,7 +299,7 @@ func (tc *E2ETestFramework) DeploySyslogReceiver(testDir string, protocol corev1
 	}
 	container := corev1.Container{
 		Name:            SyslogReceiverName,
-		Image:           "quay.io/openshift/origin-logging-rsyslog:latest",
+		Image:           "quay.io/syedriko/rsyslog:8.7-9",
```
[the remark this answers is not in the document]
Comment (syedriko) [reply test/framework/e2e/syslog.go:302]: @jcantrill I don't have permissions to create a new repository under quay/openshift-logging. This is the registry.redhat.io/rhel8/rsyslog:8.7-9 image originally.

File: internal/generator/vector/output/syslog/syslog.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	return []Element{}
+}
+
+func Facility(s *logging.Syslog) string {
+	if s == nil || s.Facility == "" {
+		return "user"
+	}
+	if IsKeyExpr(s.Facility) {
+		return fmt.Sprintf("$%s", s.Facility)
```
[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:133]: Yes. Vector configuration treats $blah as an environment variable, double $ escapes it: https://vector.dev/docs/reference/configuration/#escaping

File: internal/generator/vector/output/syslog/syslog.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	return s.Facility
+}
+
+func Severity(s *logging.Syslog) string {
+	if s == nil || s.Severity == "" {
+		return "debug"
```
[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:140]: Yep, good point. The vector side defaults to info.

File: internal/generator/vector/output/syslog/syslog.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	if IsKeyExpr(s.Severity) {
+		return fmt.Sprintf("$%s", s.Severity)
+	}
+	return s.Severity
+}
+
+func Rfc(s *logging.Syslog) string {
+	if s == nil || s.RFC == "" {
+		return "rfc5424"
```
[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:150]: I must be looking in the wrong places. This https://github.com/openshift/cluster-logging-operator/blob/9ed30e7f737b8cedbd7a69bbcfc8af6b9382443c/apis/logging/v1/output_types.go#L140 doesn't translate into anything on the Go side.

[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:150]: Added my own consts for now

File: internal/generator/vector/output/syslog/syslog.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+// Example:
+// tag
+// $.kubernetes.metadata.namespace_name
+var tagre = regexp.MustCompile(`\${tag\[-??\d+\]}`)
+
+func IsKeyExpr(str string) bool {
+	return keyre.MatchString(str)
+}
+
+func IsTagExpr(str string) bool {
```
[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:244]: Got rid of it, there're no tags in vector.

File: internal/generator/vector/output/syslog/syslog.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+// If an expression is used, its value will be taken from corresponding key in the record
+// Example: $.message.procid_key
+var keyre = regexp.MustCompile(`^\$(\.[[:word:]]*)+$`)
+
+// Example:
+// tag
+// $.kubernetes.metadata.namespace_name
+var tagre = regexp.MustCompile(`\${tag\[-??\d+\]}`)
+
+func IsKeyExpr(str string) bool {
```
[the remark this answers is not in the document]
Comment (syedriko) [reply internal/generator/vector/output/syslog/syslog.go:240]: Even better, I added vector config unit tests, that exercise this function as well.

**Comment by syedriko** [comment]
> Code looks sane overall, one missing piece: We need to set the root "level" field in our envelope to match the syslog severity. The values are equivalent, but we need to ensure we use the same capitalization and abbreviations (e.g. "warn", "WARN", "Warning") as we do elsewhere in the collector - check the fluentd syslog output and the regexps guessing levels from other log formats.

@alanconway I must be missing something. We set the .level field during normalization here https://github.com/openshift/cluster-logging-operator/blob/f8a173bfc1babe0b8de5d73cf01d37c30be8d57e/internal/generator/vector/normalize.go#L15. That's upstream and independent of the syslog sink. Are you saying the syslog sink assume .level is always there and should look at it for syslog severity? It can be done with severity = "$.message.level".

File: test/framework/e2e/syslog.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcantrill) [reply test/framework/e2e/syslog.go]: I created a quay image but am happy to defer to the redhat.io one if we expect it to be pullable by ci and not go away

[the remark this answers is not in the document]
Comment (syedriko) [reply test/framework/e2e/syslog.go]: I don't have permissions to push to quay/openshift-logging/rsyslog.

**Pull request merged**